SNAD features processor (light_curve)

VoidDruid opened the pull request:
Linked to issue(s): Closes [138](https://github.com/astrolabsoftware/fink-science/issues/138)

## If this is a new release, did you issue the corresponding schema in fink-client?

**work in progress**

## What changes were proposed in this pull request?

This PR adds a new science module with a processor that extracts required features. The processor returns an array of features for each object with length 2N (N - number of features, 2 - number of passbands/fids). Features described here - https://arxiv.org/pdf/2012.01419.pdf, package https://github.com/light-curve/light-curve-python is used for extraction.

## How was this patch tested?

In jupyter against fink test data, as described in the tutorial. Some specific tests with our test data were added (see directory tests in processor dir and file test.py), as well as doctests.

JulienPeloton commented:
Thanks @VoidDruid for the PR -- I am teaching until the end of next, so I might be slower to react. 

I enabled the workflows so that you can check the CI results. Let me know when this PR will be ready for review (without WIP).

VoidDruid commented:
@JulienPeloton I think PRs are ready for review.

JulienPeloton commented on fink_science/snad/processor.py:
```diff
@@ -0,0 +1,183 @@
+        except ValueError as err:
+            # log if known error, then skip
+            if err.args[0] == "t must be in ascending order":
+                logger.error(f"Unordered jd for {oId} in processor '{__file__}/{extract_features_snad.__name__}'")
```
I do not understand this error. The columns should be ordered by`jd` according to the line

```python
df = concat_col(df, colname, prefix=prefix)
```

And if I print one of the array, it is indeed ordered:

```
Unordered jd for ZTF18abckozp in processor '/Users/julien/Documents/workspace/myrepos/fink-science/fink_science/__init__.py/wrapper'
[2459517.7868056 2459517.7868056 2459517.7868056 2459517.8475
 2459517.8475    2459517.8475    2459517.8475    2459521.7640162
 2459521.7640162 2459522.8031366 2459530.8480324 2459531.7856481
 2459531.7856481 2459531.7856481 2459537.8261574 2459537.8261574
 2459537.8261574 2459537.8261574 2459540.7134028 2459540.7134028]
```

I think the code is complaining because there are _identical_ `jd`. Which I agree should not exist! I was looking at the data, and it seems a bug from the ZTF alert system whend doing the association between candidates  and objects. I will inspect and come back to you on this.

VoidDruid replied:
Got it, thanks. I will leave this check for now.

JulienPeloton replied:
No firm conclusion yet from ZTF -- so let's keep the code as-is.

JulienPeloton reviewed (commented):
Hi @VoidDruid, thanks for your work, this is great! I tested the code with the `run_tests.sh` script:

```bash
# in fink-science
./run_tests.sh
```

`processor.py` runs correctly. However, I was not able to run `test.py`. I left some comments for `test.py`, but there are more errors left... Could you double check this script?

VoidDruid commented on fink_science/snad/test.py:
```diff
@@ -0,0 +1,70 @@
+            split in map(lambda ln: ln.split(": "), lines)
+        })
+
+result = processor.extract_features_snad_raw(**test_params)
```
Sorry, I simply forgot to commit the updated version of this file. Everything should work now out of the box.

VoidDruid commented on fink_science/snad/test.py:
```diff
@@ -0,0 +1,70 @@
+    "633202300014898",
+    "633211400005684",
+)
+test_datasets = [pd.read_csv(f"tests/{fl}.csv") for fl in names]
```
Done

VoidDruid commented on fink_science/snad/processor.py:
```diff
@@ -0,0 +1,183 @@
+    return full_result
+
+
+extract_features_snad = udf(
```
I agree, I will do it some time in the future. I spent a lot of time trying to get it to work with pandas_udf but it's not possible with spark 2.

VoidDruid commented on fink_science/snad/fink.conf.shell:
```diff
@@ -0,0 +1,55 @@
+# Copyright 2019 AstroLab Software
```
Done

VoidDruid commented on .gitignore:
```diff
@@ -1,8 +1,10 @@
 .coverage
+.coverage*
```
Done

VoidDruid commented on README.md:
```diff
@@ -17,6 +17,7 @@ This repository contains science modules used to generate added values to alert
 | [fink_science/asteroids](fink_science/asteroids)| `roid` | int | Determine if the alert is a Solar System object |
 | [fink_science/kilonova](fink_science/kilonova)| `rf_kn_vs_nonkn` | float | probability of an alert to be a kilonova using a Random Forest Classifier (binary classification). |
 | [fink_science/nalerthist](fink_science/nalerthist)| `nalerthist` | int | Number of detections contained in each alert (current+history). Upper limits are not taken into account. |
+| [fink_science/snad](fink_science/snad)| `lc_*` | double | Numerous [light curve features](https://arxiv.org/pdf/2012.01419.pdf#section.A1) used in astrophysics. |
```
Done

JulienPeloton commented on fink_science/snad/test.py:
```diff
@@ -0,0 +1,66 @@
+features = []
+
+for name in names:
+    with open(f'tests/{name}.features') as features_file:
```
This should be:

```python
with open(f'{path}/tests/{name}.features') as features_file:
```

JulienPeloton commented:
Thanks @VoidDruid for fixing all files -- except for one remaining error, it is all fine for me! Actually, could you also merge recent change from the `master` branch (I fixed a bug in #148 ) that prevents the pipeline to run correctly? See https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork for more information.

VoidDruid commented:
> Thanks @VoidDruid for fixing all files -- except for one remaining error, it is all fine for me! Actually, could you also merge recent change from the `master` branch (I fixed a bug in #148 ) that prevents the pipeline to run correctly? See https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork for more information.

I've pulled `master` into this branch and fixed your last comment. If all looks good to you and workflows pass, this PR is ready to be merged. Thank you for reviewing!

JulienPeloton reviewed (approved):
Thanks @VoidDruid for this new science module, this is an important milestone -- I am merging it. I will perform performance tests at scale from next week.
